### PR TITLE
Three BFB mods for SCREAM

### PR DESCRIPTION
**Pull request by ambrad**
This PR implements two optimizations for SCREAM following from Noel's performance analysis:
* Noel determined that `maxval` in Homme's `compute_stage_value_dirk` is expensive. It can be replaced by `any` in two cases and moved out of the Newton iteration in another.
* Noel found that `advance_volc_radiation_data` is an expensive additional routine used in SCREAM'S DYAMOND2 configuration relative to DYAMOND1. This turned out to be for two reasons: poor index order and an O(n^2) rebin implementation. Fix both of these.

In addition, Noel pointed out that some F90-C++ interface calls in an RRM debug build were failing due to unallocated arrays. These are not used but F90 requires them to be allocated in the call. Fix this.

[BFB] for CIME E3SM and SCREAM tests; possibly non-BFB for standalone Homme theta-l tests, depending on optimization level on the machine; see comment below.

### CODE REVIEW

**Comment by ambrad** [comment]
The standalone Homme suite, non-BFB build, passes on Blake (SKX) with `-O3 -fp-model strict` but is non-BFB if using `-fp-model fast`. The changes in `imex_mod.F90` do not affect floating point arithmetic, but evidently they do induce the optimizer to change something.

**Comment by ambrad** [comment]
BFB standalone Homme cxx_vs_f90 and unit tests (enabled with `HOMMEXX_BFB_TESTING=On`) pass with `-O3 -fp-model strict` on Blake.

**Comment by ambrad** [comment]
e3sm_atm_integration e3sm_developer pass except for the currently failing hommexx test. But neither of these sets tests advance_volc_radiation_data. Does anyone know of an E3SM test that exercises that code? For now, I've used F2010-SCREAM-HR in the SCREAM repo to test that the modifications are BFB.

**Comment by ambrad** [comment]
e3sm_scream in the SCREAM repo, with this PR merged in, passes against baselines with compy_intel. I confirmed that all ten tests are sensitive to the advance_volc_radiation_data routine.

**Comment by ambrad** [comment]
@singhbalwinder I'm asking for a review from you because I modified read_volc_radiation_data.F90.

File: components/eam/src/chemistry/utils/read_volc_radiation_data.F90 [hunk 1]
```diff
@@ -381,45 +381,58 @@ subroutine advance_volc_radiation_data (specifier_sw, specifier_lw, state, pbuf2
...
-       if (read_data) then
+    if (read_data) then
+       allocate(rdata(nalts,nlats,nswbands))
+       do ifld = 1,mxnflds_sw
           !Get netcdf variable id for the field
           ierr = pio_inq_varid(piofile, trim(adjustl(specifier_sw(ifld))),var_id) !get id of the variable to read from iput file
           ! If current itp matches previous itp1, copy rather than reading from file
           if (it == itp1_old) then
-             wrk_sw(1,ifld,:,:,:) = wrk_sw(2,ifld,:,:,:)
+             wrk_sw(:,1,:,:,ifld) = wrk_sw(:,2,:,:,ifld)
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/utils/read_volc_radiation_data.F90:391]: Thanks for fixing this index ordering. This should make a difference.

**Review by singhbalwinder — APPROVED** [review]
Thanks @ambrad and @ndkeen for fixing these issues. I have some minor questions/comments but otherwise it looks great!

File: components/eam/src/chemistry/utils/read_volc_radiation_data.F90 [hunk 2]
```diff
@@ -381,45 +381,58 @@ subroutine advance_volc_radiation_data (specifier_sw, specifier_lw, state, pbuf2
 
     fact2 = 1._r8-fact1
 
-    do ifld = 1,mxnflds_sw
-       if (read_data) then
+    if (read_data) then
+       allocate(rdata(nalts,nlats,nswbands))
```
[the remark this answers is not in the document]
Comment (ambrad) [reply components/eam/src/chemistry/utils/read_volc_radiation_data.F90:385]: Keep in mind this part of the code runs once per simulated month, not once per time step, and it's calling I/O routines, which themselves will likely do a lot of de/allocation. In the low-res case, these de/alllocs happen once per ~1500 time steps. Thus, doing two de/allocations (one here and one for lw) is inconsequential. Better to do that than leave rarely used arrays hanging around, IMO.

[the remark this answers is not in the document]
Comment (singhbalwinder) [reply components/eam/src/chemistry/utils/read_volc_radiation_data.F90:385]: Agreed, If these are used rarely then it should be fine to keep these as it is.

**Review by ndkeen — APPROVED** [review]

**Comment by oksanaguba** [comment]
in next

**Comment by rljacob** [comment]
No impact on v2.  Can be merged.

**Pull request merged**